bluesky/simulators.py

```diff
@@ -114,3 +114,276 @@ def check_limits(plan):
+    on statistics gathered from previous use, and returns the mean value (est_time) and the 
+    standard deviation (std_dev) from those statistics, est_time also returns the std_dev for the 
+    est_time to give an idea of the accuracy of the prediction. As this is a statistical approach 
+    to time estimating at the device level it is expected that the accuracy will improve with more 
+    use of the device through bluesky. The final information returned is a list of est_time, std_dev
+    pairs for each 'run' in the plan, where a 'run' is defined as the anything that occurs between 
+    a 'start' and 'stop' document being generated (which is how the data is stored in the 
+    databroker). 
+
+    Parameters
+    ----------
+    plan : generator.
+        The bluesky plan that the est_time is to be estimated for.
+    print_output : boolean, optional.
+        Indicates if the return values should also be printed to the command line in a human 
+        readable way, default value is 'True'.
+    
+    Return Parameters
+    -----------------
+    out_est_time : list.
+        A list containing 2 items, the est_time and the std_dev, for the plan.
+    run_info : list.
+        A list of items, 1 for each run in the plan, containing an est_time/std_dev tuple.
+
+    """
+
+    #The commands that start/end a 'run'.
+    _run_start_cmds = ['open_run']
+    _run_end_cmds = ['close_run']
+    #The commands that start/end a 'group', but not flyer commands.
+    _group_start_cmds = ['set','trigger']   
+    _group_end_cmds = ['wait']
+    #The commands that don't start/end a 'group/plan' but are still timed.
+    _timed_cmds = ['stage', 'unstage', 'read', 'sleep']
+    #Commands that are 
+    _flyer_start_cmds = ['kickoff']
+
+
+
+    def combine_est_time(est_time_1, est_time_2, method = 'sum'):
+        """
+        Returns the combination est_time/std_dev pairs et_1 and et_2.
+        This function returns the combination of est_time_1 and est_time_2, combined using the 
+        method definedby 'method'.
+
+        Parameters
+        ----------
+        est_time_1, est_time_2 : tuples.
+            The tuples containing the est_time/std_dev tuples to be combined.
+        method : string, optional.
+            The method to use for the combination of est_time_1 and est_time_2, default is 
+            'sum'.
+
+        Return Parameters
+        -----------------
+        out_est_time : list.
+            The combined est_time/std_dev tuple.
+        """
+        out_est_time = [0,0]
+        if method == 'sum':
+            out_est_time[0] = est_time_1[0] + est_time_2[0]
+            out_est_time[1] = est_time_1[1] + est_time_2[1]
+
+        elif method == 'max':
+            if est_time_1[0] < est_time_2[0]:
+                out_est_time = est_time_2 
+
+            else:
+                out_est_time = est_time_1
+
+        return out_est_time
+
+
+
+    def obj_est_time(msg, val_dict):
+        """
```

Comment:
Functions that mutate their inputs should document prominent that they do so. It can be useful, and seems useful in the case, but it can also be _very surprising_ so it needs to be clearly flagged.

Reply:
I thought I had clearly flagged this, I will make it clearer in the doc string.

Reply:
Not sure exactly how to do this as a class, but I will think about it. Is the idea that an instance of the 'class' will have the dictionary as an attribute?